### PR TITLE
chapter_10

### PR DESCRIPTION
**Pull request by lrmi0808**
Ready

### CODE REVIEW

**Comment by lrmi0808** [comment]
Ready

File: app/helpers/sessions_helper.rb [hunk 1]
```diff
@@ -36,4 +32,21 @@ def forget user
     cookies.delete(:user_id)
     cookies.delete(:remember_token)
   end
+
+  def logged_in?
+    !current_user.nil?
```
Comment (ngocvt-0484) [inline app/helpers/sessions_helper.rb:37]:
```suggestion
    current_user.present?
```

File: app/views/users/_user.html.erb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+  <li>
+    <%= gravatar_for user, size: 50 %>
+    <%= link_to user.name, user %>
+    <% if current_user.admin? && !current_user?(user) %> | <%= link_to t("delete"), user, method: :delete, data: {confirm: t("sure")} %>
+<% end %>
```
Comment (ngocvt-0484) [inline app/views/users/_user.html.erb:5]: wrong indent

File: app/views/users/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<% provide(:title, t("all_users")) %>
+<h1><%= t("all_users") %></h1>
+
+<%= will_paginate @users %>
+
+<ul class="users">
+  <% @users.each do |user| %>
```
Comment (ngocvt-0484) [inline app/views/users/index.html.erb:7]: just only need 
```suggestion
 <%= render @users %>
```
you can remove 
```
  <% @users.each do |user| %>
```

**Comment by lrmi0808** [comment]
Ready

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,4 +1,12 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:2]:
```suggestion
  before_action :logged_in_user, except: %i(new show create)
```

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,4 +1,12 @@
...
+  before_action :correct_user, only: [:edit, :update]
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:3]:
```suggestion
  before_action :correct_user, only: %i(edit update)
```

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by lrmi0808** [comment]
Ready

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -23,9 +31,54 @@ def create
...
+  end
+
+  def destroy
+    user = User.find_by(id: params[:id])
+    if user&.destroy
+    flash[:success] = t("user_deleted")
+    else
+    flash[:danger] = t("delete_fail")
+    end
+    redirect_to users_url
```
Comment (anlvq-1058) [inline app/controllers/users_controller.rb:57]: please ident

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by lrmi0808** [comment]
Ready

File: config/routes.rb [hunk 7]
```diff
@@ -6,10 +6,15 @@
...
       get "/signup", to: "users#new"
       post "/signup", to: "users#create"
-      resources :users, only: %i(show new create destroy)
+      resources :users, only: %i(show new create destroy edit update)
 
       get "/login", to: "sessions#new"
       post "/login", to: "sessions#create"
       delete "/logout", to: "sessions#destroy"
+
+      get "/edit", to: "users#edit"
```
Comment (huongnt-2545) [inline config/routes.rb:15]: remove line 15 -> 18, because these resources already at line 9

File: config/routes.rb [hunk 8]
```diff
@@ -6,10 +6,15 @@
 
       get "/signup", to: "users#new"
       post "/signup", to: "users#create"
-      resources :users, only: %i(show new create destroy)
+      resources :users, only: %i(show new create destroy edit update)
```
Comment (huongnt-2545) [inline config/routes.rb:9]: comprehensive actions, you should define: `resources :users`

File: app/models/user.rb [hunk 9]
```diff
@@ -40,10 +37,13 @@ def remember
...
   def authenticated? remember_token
+    return false if remember_digest.nil?
     BCrypt::Password.new(remember_digest).is_password? remember_token
   end
 
   def forget
     update_attribute :remember_digest, nil
   end
+
+  private
```
Comment (huongnt-2545) [inline app/models/user.rb:48]: remove this line

File: app/models/user.rb [hunk 10]
```diff
@@ -3,18 +3,15 @@ class User < ApplicationRecord
...
+  validates :password, presence: true, length: {minimum: 6}, allow_nil: true
 
   has_secure_password
 
   #email lowercase  conversion
   before_save :downcase_email
 
-  private
-
   def downcase_email
```
Comment (huongnt-2545) [inline app/models/user.rb:15]: move this method place to the end of class, and set private scope
```
private
def downcase_email
  self.email.downcase!
end
```

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -23,9 +31,54 @@ def create
     end
   end
 
+
+
+  def edit
+    @user = User.find_by(id: params[:id])
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb:37]: remove this line, it's no need

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -23,9 +31,54 @@ def create
...
+  end
+
+  def update
+    @user = User.find_by(id: params[:id])
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb:41]: remove this line, it;s no need

**Comment by lrmi0808** [comment]
Ready

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (anlvq-1058) [reply app/controllers/users_controller.rb:35]: please fix this

**Review by anlvq-1058 — COMMENTED** [review]
cmt

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (lrmi0808) [reply app/controllers/users_controller.rb:35]: > please fix this

image

This error occurs if I delete that line. Can I know how to solve this?

**Comment by lrmi0808** [comment]
Ready

**Review by anlvq-1058 — COMMENTED** [review]
merged

**Pull request merged**